src/backend/distributed/planner/recursive_planning.c

```diff
@@ -1337,6 +1348,211 @@ NodeContainsSubqueryReferencingOuterQuery(Node *node)
+	subquery->jointree->quals = (Node *) andedBoundExpressions;
+
+	/* force recursively planning of the newly created subquery */
+	subquery->limitOffset = (Node *) MakeIntegerConst(0);
+
+	/* replace the function with the constructed subquery */
+	rangeTableEntry->rtekind = RTE_SUBQUERY;
+	rangeTableEntry->subquery = subquery;
+
+	/*
+	 * If the relation is inherited, it'll still be inherited as
+	 * we've copied it earlier. This is to prevent the newly created
+	 * subquery being treated as inherited.
+	 */
+	rangeTableEntry->inh = false;
+
+	if (IsLoggableLevel(DEBUG1))
+	{
+		StringInfo subqueryString = makeStringInfo();
+
+		pg_get_query_def(subquery, subqueryString);
+
+		ereport(DEBUG1, (errmsg("Wrapping local relation \"%s\" to a subquery: %s ",
+								get_rel_name(rangeTableEntry->relid),
+								ApplyLogRedaction(subqueryString->data))));
+	}
+	RecursivelyPlanSubquery(rangeTableEntry->subquery, context);
+}
+
+
+/*
+ * ContainsTableToBeConvertedToSubquery checks if the given range table list contains
+ * any table that should be converted to a subquery, which otherwise is not plannable.
+ */
+bool
+ContainsTableToBeConvertedToSubquery(List *rangeTableList, Oid resultRelationId)
+{
+	if (AllDataLocallyAccessible(rangeTableList))
+	{
+		return false;
+	}
+	if (ContainsLocalTableDistributedTableJoin(rangeTableList))
+	{
+		return true;
+	}
+	if (ContainsLocalTableSubqueryJoin(rangeTableList, resultRelationId))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * AllDataLocallyAccessible return true if all data for the relations in the
+ * rangeTableList is locally accessible.
+ */
+static bool
+AllDataLocallyAccessible(List *rangeTableList)
+{
+	RangeTblEntry *rangeTableEntry = NULL;
+	foreach_ptr(rangeTableEntry, rangeTableList)
+	{
+		if (rangeTableEntry->rtekind == RTE_SUBQUERY)
+		{
+			/* TODO:: check if it has distributed table */
+			return false;
+		}
+		if (!IsRecursivelyPlannableRelation(rangeTableEntry))
+		{
+			continue;
+		}
+
+
+		Oid relationId = rangeTableEntry->relid;
+
+		if (!IsCitusTable(relationId))
+		{
+			/* local tables are locally accessible */
+			continue;
+		}
+
+		List *shardIntervalList = LoadShardIntervalList(relationId);
+		if (list_length(shardIntervalList) != 1)
+		{
+			/* we currently only consider single placement tables */
+			return false;
+		}
+
+		ShardInterval *shardInterval = linitial(shardIntervalList);
+		uint64 shardId = shardInterval->shardId;
+		ShardPlacement *localShardPlacement =
+			ShardPlacementOnGroup(shardId, GetLocalGroupId());
+		if (localShardPlacement == NULL)
+		{
+			/* the table doesn't have a placement on this node */
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * IsRecursivelyPlannableRelation returns true if the given range table entry
+ * is a relation type that can be converted to a subquery.
+ */
+bool
+IsRecursivelyPlannableRelation(RangeTblEntry *rangeTableEntry)
+{
+	if (rangeTableEntry->rtekind != RTE_RELATION)
+	{
+		return false;
+	}
```

Comment:
here we should probably consider some other things as well, like FOREIGN TABLE, MAT VIEW

Reply:
@onderkalaci any other relkind we should have here?

Reply:
these look good